pydis_core/exts/source.py

```diff
@@ -0,0 +1,198 @@
+
+from discord import Embed
+from discord.ext import commands
+from discord.utils import escape_markdown
+
+if TYPE_CHECKING:
+    from pydis_core import BotBase as Bot
+
+
+GITHUB_AVATAR = "https://avatars1.githubusercontent.com/u/9919"
+BOT_CORE_REPO = "https://github.com/python-discord/bot-core"
+
+class _TagIdentifierStub(NamedTuple):
+    """A minmally functioning stub representing a tag identifier."""
+
+    group: str | None
+    name: str
+
+    @classmethod
+    def from_string(cls, string: str) -> "_TagIdentifierStub":
+        """Create a TagIdentifierStub from a string."""
+        split_string = string.split(" ", maxsplit=2)
+        if len(split_string) == 1:
+            return cls(None, split_string[0])
+        return cls(split_string[0], split_string[1])
+
+
+class _SourceType(enum.StrEnum):
+    """The types of source objects recognized by the source command."""
+
+    help_command = enum.auto()
+    text_command = enum.auto()
+    core_command = enum.auto()
+    cog = enum.auto()
+    core_cog = enum.auto()
+    tag = enum.auto()
+    extension_not_loaded = enum.auto()
+
+
+class SourceCode(commands.Cog, description="Displays information about the bot's source code."):
+    """
+    Pre-built cog to display source code links for commands and cogs (and if applicable, tags).
+
+    To use this cog, instantiate it with the bot instance and the base GitHub repository URL.
+
+    Args:
+        bot (:obj:`pydis_core.BotBase`): The bot instance.
+        github_repo: The base URL to the GitHub repository (e.g. `https://github.com/python-discord/bot`).
+    """
+
+    def __init__(self, bot: "Bot", github_repo: str) -> None:
+        self.bot = bot
+        self.github_repo = github_repo
+
+    @commands.command(name="source", aliases=("src",))
+    async def source_command(
+        self,
+        ctx: commands.Context["Bot"],
+        *,
+        source_item: str | None = None,
+    ) -> None:
+        """Display information and a GitHub link to the source code of a command, tag, or cog."""
+        if not source_item:
+            embed = Embed(title=f"{self.bot.user.name}'s GitHub Repository")
+            embed.add_field(name="Repository", value=f"[Go to GitHub]({self.github_repo})")
+            embed.set_thumbnail(url=GITHUB_AVATAR)
+            await ctx.send(embed=embed)
+            return
+
+        obj, source_type = await self._get_source_object(ctx, source_item)
+        embed = await self._build_embed(obj, source_type)
+        await ctx.send(embed=embed)
+
+    @staticmethod
+    async def _get_source_object(ctx: commands.Context["Bot"], argument: str) -> tuple[object, _SourceType]:
+        """Convert argument into the source object and source type."""
+        if argument.lower() == "help":
+            return ctx.bot.help_command, _SourceType.help_command
+
+        cog = ctx.bot.get_cog(argument)
+        if cog:
+            if cog.__module__.startswith("pydis_core.exts"):
```

Comment:
what about using `inspect.getmodule()`? It *should* be less fragile.

Reply:
43c8d2b61f97ac00fccda1806d345791bb3ea10e